orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py

```diff
@@ -0,0 +1,194 @@
+                                    keep_initializers_as_inputs=True,
+                                    training=torch.onnx.TrainingMode.TRAINING)
+                        except Exception as e:
+                            # If this module is not exportable for one arg
+                            # group, we say this module is not exportable.
+                            module_exportable = False
+                            # Already found a broken case.
+                            # No need to check next case.
+                            break
+
+                    exportable_list[module] = module_exportable
+                    return exportable_list[module]
+
+        # Add a hook to record forward's input for all modules.
+        recursive_hook(self._original_module)
+
+        # Run forward with actual input to record all possible
+        # inputs for all invoked modules.
+        y = self._original_module(*args, **kwargs)
+
+        # We already have "supported_modules" so
+        # we no longer need those hooks in forward pass.
+        for h in handle_pool:
+            h.remove()
+
+        # Try exporter on all module-input pairs. If a module can be exported with
+        # all its recorded inputs, then it's exporable.
+        check_exportable(self._original_module)
+
+        # A naive way of determining if ORT can run nn.Module
+        def is_supported(module):
+            return module in exportable_list and exportable_list[module]
+
+        # Top-down wrapper to replace nn.Module's with ORTModule.
+        # Note that using bottom-up wrapper may lead to much
+        # ORTModule instances and each ORTModule owns a much smaller graph.
+        def recursive_wrap(module):
+            sub_dict = module._modules
+            for name, sub in sub_dict.items():
+                if isinstance(sub, torch.nn.ModuleList):
+                    # We encounter a list of sub-modules.
+                    # Let's wrap them one-by-one.
+                    for name1, sub1 in sub._modules.items():
+                        if is_supported(sub1):
+                            sub._modules[name1] = ORTModule(
+                                sub1,
+                                debug_options=self._debug_options)
+                        else:
+                            recursive_wrap(sub1)
+                else:
+                    if is_supported(sub):
+                        # Just wrap it as ORTModule when possible.
+                        sub_dict[name] = ORTModule(
+                            sub,
+                            debug_options=self._debug_options)
+                    else:
+                        # This sub-module is not exportable to ONNX
+                        # Let's check its sub-modules.
+                        recursive_wrap(sub)
+
+        recursive_wrap(self._original_module)
+
+    def forward(self, *inputs, **kwargs):
+        if not self._initialized:
+            self._initialized = True
```

Comment:
Shall we mark `self._initialized = True` after completing `self._initialize(...)` so that we prevent unexpected results if `self._initialize(...)` is unsuccessful and users call `forward` again?

Reply:
Sure. We set `self._initialized=True` at the end of `self._initialize(...)`.